### PR TITLE
Feature: complex field grammar

### LINKED ISSUE #50: Parsing equals for object
**Describe the bug**

Update grammare to permit to use = for object.

this code not work with actual parser: 
```tf
provider "aws" {
  test = {
    value = 1
  }
}
```

### PR DESCRIPTION
**Pull request by MilannGuitton**
Closed #50

### CODE REVIEW

File: src/models/TerraformComponentAttribute.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import { ComponentAttribute } from 'leto-modelizer-plugin-core';
+
+/**
+ * Specific Terraform component attribute.
+ *
+ * @augments {ComponentAttribute}
+ */
+class TerraformComponentAttribute extends ComponentAttribute {
```
Comment (Zorin95670) [inline src/models/TerraformComponentAttribute.js:8]: missing unit test: `tests/unit/models/TerraformComponentAttribute.spec.js`

File: src/parser/TerraformListener.js [hunk 2]
```diff
@@ -226,40 +226,79 @@ class TerraformListener extends antlr4.tree.ParseTreeListener {
         this.currentField.name,
       );
     }
-
     this.currentField = null;
   }
 
   // Enter a parse tree produced by terraformParser#complexField.
-  enterComplexField(ctx) {
+  enterComplexField() {
```
Comment (Zorin95670) [inline src/parser/TerraformListener.js:233]: I'm not sure, but complexe field are remove in grammar, so you can delete related functions

File: src/parser/TerraformListener.js [hunk 3]
```diff
@@ -226,40 +226,79 @@ class TerraformListener extends antlr4.tree.ParseTreeListener {
...
       value: [],
       type: 'Object',
       definition,
     });
   }
 
-  // Exit a parse tree produced by terraformParser#complexField.
-  exitComplexField() {
+  // Exit a parse tree produced by terraformParser#attributeBlock.
+  exitAttributeBlock() {
```
Comment (Zorin95670) [inline src/parser/TerraformListener.js:293]: same content as exitDynamicBlock, please add a function `addObjectField()` like `addComponent()`

File: src/parser/TerraformListener.js [hunk 4]
```diff
@@ -226,40 +226,79 @@ class TerraformListener extends antlr4.tree.ParseTreeListener {
...
+    });
+  }
+
+  // Exit a parse tree produced by terraformParser#dynamicBlock.
+  exitDynamicBlock() {
+    if (this.fieldsTree.length > 0) {
+      const field = this.fieldsTree.pop();
+
+      field.value.push(this.currentComplexField);
+      this.currentComplexField = field;
```
Comment (Zorin95670) [inline src/parser/TerraformListener.js:266]: complexeField no more exist in grammar, so can you rename this variable in `currentObjectField`

File: scripts/generate_parser.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
+const fs = require('fs');
+const { exec } = require('node:child_process');
+
+new Promise((resolve) => {
+  exec('java -Xmx500M -cp "/usr/local/lib/antlr-4.11.0-complete.jar" org.antlr.v4.Tool -Dlanguage=JavaScript src/antlr/terraformParser.g4 -lib src/antlr/', (error, stdout, stderr) => {
```
Comment (Zorin95670) [inline scripts/generate_parser.js:5]: not working on windows

File: README.md [hunk 6]
```diff
@@ -10,6 +10,16 @@ npm run build
 
 ## Development
 
+### Generate the parser
```
Comment (Zorin95670) [inline README.md:13]: missing update of changelog

File: src/models/TerraformComponentAttribute.js (hunk 1 above)
Comment (MilannGuitton) [reply src/models/TerraformComponentAttribute.js:8]: Done

File: src/parser/TerraformListener.js (hunk 2 above)
Comment (MilannGuitton) [reply src/parser/TerraformListener.js]: Done

File: src/parser/TerraformListener.js (hunk 3 above)
Comment (MilannGuitton) [reply src/parser/TerraformListener.js:279]: Done

File: README.md (hunk 6 above)
Comment (MilannGuitton) [reply README.md:13]: Done

File: src/parser/TerraformListener.js (hunk 4 above)
Comment (MilannGuitton) [reply src/parser/TerraformListener.js]: Done

File: changelog.md [hunk 7]
```diff
@@ -11,6 +11,17 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - Update plugin-core to version 0.14.0
 
+## [0.1.13] - 2023/03/16
```
Comment (Zorin95670) [inline changelog.md:14]: It's not a release pull request, put content in [Unreleased] section

File: src/antlr/terraformParser.g4 [hunk 8]
```diff
@@ -7,7 +7,7 @@ file
   ;
 
 dataDirective
-  : DATA providerType name AO (object)+ AF
+  : DATA providerType name AO object* AF
```
Comment (Zorin95670) [inline src/antlr/terraformParser.g4:10]: why not for other terraform blocks (provider, module, terraform, variable) ?

**Review by Zorin95670 — CHANGES_REQUESTED** [review]

File: src/antlr/terraformParser.g4 (hunk 8 above)
Comment (MilannGuitton) [reply src/antlr/terraformParser.g4:10]: The `terraform validate` command throws an error when no `source` argument is provided to a module block,or no `value` argument is provided to an output block. However,  the other directives can indeed have an empty body. I changed it accordingly.

File: changelog.md (hunk 7 above)
Comment (MilannGuitton) [reply changelog.md]: Done

File: src/antlr/terraformParser.g4 [hunk 9]
```diff
@@ -55,16 +55,19 @@ type
...
 
 field
   : (IDENTIFIER|SOURCE) EQUAL expression
   ;
 
-complexField
-  : IDENTIFIER AO object AF
-  | IDENTIFIER EQUAL AO (STRING EQUAL STRING)+ AF
+attributeBlock
+  : IDENTIFIER EQUAL AO object* AF
```
[the remark this answers is not in the document]
Comment (MilannGuitton) [reply src/antlr/terraformParser.g4:66]: The `object*` expression should already match 0 or more objects, thus allowing an empty block body.

**Pull request closed without merging**